### DISCUSSION TITLE
How does OC handle c# decimal or money types (json -> frontend)

### DISCUSSION THREAD

**Discussion by bashuss** [OrchardCMS/OrchardCore#16211] [Q&A]
As OC stores data as json, there is basically no difference in storage between decimal, float and double. So the difference is made by the underlying model the json is deserialized to. 
Is this correct?

But if I am using queries from db and use these in liquid templates, or I provide the json to the frontend, how is this to be handled properly?

To keep the precision of a decimal, frontend javascript would need to consume the json as a string and not as a number. This string could then be displayed as string, or assigned to a js implementation of decimal (I have found, but not tested this: https://github.com/MikeMcl/decimal.js).

Has there been put any thought into this already? 

As a background for my question, you may check this:
https://stackoverflow.com/questions/3730019/why-not-use-double-or-float-to-represent-currency

**Comment by Piedone** [comment]
What exactly is your use case? Storing such values in content items?

With e.g. a `decimal` property of a content part, it'll be serialized to JSON with `System.Text.Json`. Even though the representing string of it might be the same as that of a `float` or `int`, for example, during deserialization you'll still get a `decimal` back. There shouldn't be any data loss.

**Reply by bashuss** [reply]
The use case is, that I am handling data (money and money calculations), which needs to have exact decimal representation (therefore one solution is the decimal type in c#). But as I may need to use the values in the frontend and may even need to do small calculations in the frontend, these numbers need to be provided to the frontend javascript. 
My question now is, if there is always a stage of deserialization the database stored json to a c# model, before the data is presented to the frontend (be it coupled or decoupled via api). 
Are there any cases I need to be aware of, in which automatically a json number is passed to the frontend, or a number type with less precision is used (maybe in liquid templates?).

My best guesses for possible problems would be:

1. backend queries used in templates
2. liquid templates
3. any places where we directly pass the json of db to frontend (is there any standard scenario or would that always be bad individual design?)
4. Api providing json without adjusted serializer
5. Internal OC javascript engine

If some of these are actually possible problems, what would be the best way to address these on the OC side.

Is this clarifying my question a bit?

**Reply by Piedone** [reply]
I don't know about what would happen in Liquid, but for views, I'd definitely recommend using Razor. At that point, you'll have the same types everywhere for sure.

As for all the other cases, including Liquid in something like query templates defined on the admin, I don't know. I suggest you do a couple of tests to see.

**Reply by bashuss** [reply]
Seems like 2. could be ok: https://github.com/sebastienros/fluid/blob/main/Fluid/Values/NumberValue.cs
NumberValue is internally a c# decimal. Also OC number fields are decimal.
